### PR TITLE
Broken links in tracking results

### PR DESCRIPTION
**Pull request by shershennm**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | x
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | #3599
| BC breaks? | x
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Go to page with ampersands in url
2. Url will be stored as test.com/test?first_name=John&#38;last_name=Smith&#38;utm_campaign=test

#### Steps to test this PR:
1. Go to page with ampersands in url
2. Url must be stored without it breaking and clean from xss injections

# BC Breaks
Solved by https://github.com/mautic/mautic/pull/6712

### CODE REVIEW

**Comment by escopecz** [comment]
Can you please write unit tests for the new method?

**Comment by escopecz** [comment]
I find the steps to test hard to follow. I don't know how accessing a URL with ampersands is related to Mautic. So I tried to follow the steps in the linked issue, but I cannot find a way how to get the ampersands as url-encoded alternative. I tried to reproduce via a page hit query param, but the URL is saved correctly with ampersands. I checked in the database to be sure. Then I tried to send the URL as a custom field in an email and it was send with ampersands too. I even checked the email source code to confirm.

Please provide more stupid-proof steps to replicate.

**Comment by shershennm** [comment]
The problem is in Mautic tracking. When the user goes to the page with ampersands in URL, this URL in "Page hit" event will be incorrectly encoded. It also will be incorrect in database. 
I found mentioned issue similar, but it about another (solved?) problem. Sorry for wrong information

**Comment by escopecz** [comment]
I still don't understand. Please provide detailed steps to reproduce that the testers can follow.

**Comment by shershennm** [comment]
1. Install Mautic tracking code on a website
2. Go to this website to page with ampersands in URL
3. Watch this page view in Mautic - the URL in "Page Hit" event will be wrong
screenshot from 2018-05-21 22-02-25

**Comment by npracht** [comment]
@escopecz 
# Description:
If the URL has ampersands, they will not match campaign page hit decisions because the URL is encoded prior to storing making comparisons not match.

## Steps to reproduce the bug:
Create a template email or SMS with a link that has at least one ampersand &
Create a campaign to send the message with a page hits decision that matches the full URL
Execute the campaign then click on the link
The campaign decision is not logged

## Steps to test this PR:
Repeat and the campaign decision will be logged

**Comment by escopecz** [comment]
I don't see any problem in the code. Migration would be welcome. It will update records in the stat files which can go to millions of rows in some Mautic instances so it can cause troubles during upgrade. But I hope that admins of those big instances don't use the web upgrade feature but rather the upgrade script.

I'd suggest to use query in the migration similar to this:
```
UPDATE `page_hits`
SET `custom_html` = replace(custom_html, '&#38;', '&')
```
_(not tested)_

**Comment by npracht** [comment]
@shershennm you can manage that ?

**Comment by escopecz** [comment]
Ok, so I tested it on page_table with 2.33M rows. It affected 468K rows and took 23.8 seconds. Now we have to decide if it's fast enough. Do we need it for referer too?

**Comment by npracht** [comment]
i'd say wherever you can filter on a url. so yes ?
Would you add this in a commit @escopecz ? I'm not sure we can have a fast feedback from @shershennm as far as the PR is a bit old.

**Comment by Gregy** [comment]
I don't think a query like this will work on tables with tens of millions of rows. If you want to make it work for the really large tables you would have to:
1. select only the rows that contain '&#38;' and store their ids (to avoid searching for them each iteration)
2. loop the update statement and limit it to ~30000 ids at a time (to avoid too large transaction)
